### PR TITLE
Add unit tests for GenerateSbomTask arguments

### PR DESCRIPTION
**Pull request by vpatakottu**
Added unit tests to validate arguments passed into GenerateSbomTask. Also added root checks for all the args involving paths.

### CODE REVIEW

File: src/Microsoft.Sbom.Targets/GenerateSbomTask.cs [hunk 1]
```diff
@@ -121,6 +122,22 @@ public override bool Execute()
...
+                (!string.IsNullOrEmpty(this.BuildComponentPath) && !Path.IsPathRooted(this.BuildComponentPath)) ||
+                (!string.IsNullOrEmpty(this.ManifestDirPath) && !Path.IsPathRooted(this.ManifestDirPath)) ||
+                (!string.IsNullOrEmpty(this.ExternalDocumentListFile) && !Path.IsPathRooted(this.ExternalDocumentListFile)))
+            {
+                Log.LogError("SBOM generation failed: Unrooted path detected. Please ensure you are specifying full paths for " +
+                    "BuildDropPath, BuildComponentPath, ManifestDirPath, and ExternalDocumentListFile arguments.");
+                return false;
+            }
+
+            if (!(string.IsNullOrEmpty(NamespaceUriUniquePart) || Guid.TryParse(NamespaceUriUniquePart, out var guidResult))) {
```
Comment (gustavoaca1997) [inline src/Microsoft.Sbom.Targets/GenerateSbomTask.cs]: For more readability, I would write it like this:
```suggestion
            if (!string.IsNullOrEmpty(NamespaceUriUniquePart) && !Guid.TryParse(NamespaceUriUniquePart, out var guidResult)) {
```

File: src/Microsoft.Sbom.Targets/GenerateSbomTask.cs [hunk 2]
```diff
@@ -121,6 +122,22 @@ public override bool Execute()
     {
         try
         {
+            // Check if all arguments involving paths are rooted
+            if (!Path.IsPathRooted(this.BuildDropPath) ||
+                (!string.IsNullOrEmpty(this.BuildComponentPath) && !Path.IsPathRooted(this.BuildComponentPath)) ||
```
Comment (gustavoaca1997) [inline src/Microsoft.Sbom.Targets/GenerateSbomTask.cs]: Can you change all the `IsNullOrEmpty` to `IsNullOrWhiteSpace`?

File: test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        yield return new object[] { CurrentDirectory, PackageSupplier, PackageName, PackageVersion, null };
+    }
+
+    private static IEnumerable<object[]> GetEmptyRequiredParamsData()
+    {
+        yield return new object[] { string.Empty, PackageSupplier, PackageName, PackageVersion, NamespaceBaseUri };
+        yield return new object[] { CurrentDirectory, string.Empty, PackageName, PackageVersion, NamespaceBaseUri };
+        yield return new object[] { CurrentDirectory, PackageSupplier, string.Empty, PackageVersion, NamespaceBaseUri };
+        yield return new object[] { CurrentDirectory, PackageSupplier, PackageName, string.Empty, NamespaceBaseUri };
+        yield return new object[] { CurrentDirectory, PackageSupplier, PackageName, PackageVersion, string.Empty };
```
Comment (gustavoaca1997) [inline test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs:107]: Can you add cases with empty spaces, tabs and new lines?

File: test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    public void Sbom_Fails_With_Invalid_NamespaceBaseUri()
+    {
+        // Arrange
+        var task = new GenerateSbomTask
+        {
+            BuildDropPath = CurrentDirectory,
+            PackageSupplier = PackageSupplier,
+            PackageName = PackageName,
+            PackageVersion = PackageVersion,
+            NamespaceBaseUri = "incorrectly_formatted_uri.com",
```
Comment (gustavoaca1997) [inline test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs]: Can you add more cases, such as URIs with invalid characters in the middle, spaces, new lines, etc?
Maybe we should add either here or in the tests I added different cases for valid URIs, such as having uppercase and lowercase character combined. We can create a deliverable for those extra cases.

File: test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    {
+        // Arrange
+        var task = new GenerateSbomTask
+        {
+            BuildDropPath = CurrentDirectory,
+            PackageSupplier = PackageSupplier,
+            PackageName = PackageName,
+            PackageVersion = PackageVersion,
+            NamespaceBaseUri = NamespaceBaseUri,
+            NamespaceUriUniquePart = "-1",
```
Comment (gustavoaca1997) [inline test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs]: Can you add extra cases for invalid values of the unique part? Such as too long or too short strings, with invalid characters, spaces, new lines, tabs, etc?

File: src/Microsoft.Sbom.Targets/GenerateSbomTask.cs [hunk 6]
```diff
@@ -121,6 +122,22 @@ public override bool Execute()
...
+                (!string.IsNullOrEmpty(this.ManifestDirPath) && !Path.IsPathRooted(this.ManifestDirPath)) ||
+                (!string.IsNullOrEmpty(this.ExternalDocumentListFile) && !Path.IsPathRooted(this.ExternalDocumentListFile)))
+            {
+                Log.LogError("SBOM generation failed: Unrooted path detected. Please ensure you are specifying full paths for " +
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/GenerateSbomTask.cs]: Suggest putting what path is not rooted and splitting up the if check so you can provide very specific logging when they have a wrong path.

nameof(this.BuildComponentPath), this.BuildComponentPath

File: src/Microsoft.Sbom.Targets/GenerateSbomTask.cs [hunk 7]
```diff
@@ -121,6 +122,22 @@ public override bool Execute()
...
+                    "BuildDropPath, BuildComponentPath, ManifestDirPath, and ExternalDocumentListFile arguments.");
+                return false;
+            }
+
+            if (!(string.IsNullOrEmpty(NamespaceUriUniquePart) || Guid.TryParse(NamespaceUriUniquePart, out var guidResult))) {
+                Log.LogError("SBOM generation failed: NamespaceUriUniquePart must be a valid GUID");
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/GenerateSbomTask.cs]: Would print out what the NamespaceUriUniquePart was

Suggestion
Log.LogError($"SBOM generation failed: NamespaceUriUniquePart '{NamespaceUriUniquePart}' must be a valid GUID");

The $ formatted string will handle null here and not break.

File: test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+        // Act
+        var result = task.Execute();
+
+        // Assert
+        Assert.IsFalse(result);
+    }
+
+    private static IEnumerable<object[]> GetUnrootedPathTestData()
+    {
+        yield return new object[] { "../../", BuildComponentPath, DefaultManifestDirectory, ExternalDocumentListFile };
```
Comment (gustavoaca1997) [inline test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs]:
```suggestion
        yield return new object[] { Path.Combine("..", ".."), BuildComponentPath, DefaultManifestDirectory, ExternalDocumentListFile };
```

File: src/Microsoft.Sbom.Targets/GenerateSbomTask.cs (hunk 6 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/GenerateSbomTask.cs]: Updated!

File: src/Microsoft.Sbom.Targets/GenerateSbomTask.cs (hunk 7 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/GenerateSbomTask.cs]: Updated!

File: src/Microsoft.Sbom.Targets/GenerateSbomTask.cs (hunk 2 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/GenerateSbomTask.cs]: Updated!

File: src/Microsoft.Sbom.Targets/GenerateSbomTask.cs (hunk 1 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/GenerateSbomTask.cs]: Updated!

File: test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs (hunk 3 above)
Comment (vpatakottu) [reply test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs:107]: updated!

File: test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs (hunk 4 above)
Comment (vpatakottu) [reply test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs]: Added!

File: test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs (hunk 5 above)
Comment (vpatakottu) [reply test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs]: Added!

File: test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs (hunk 8 above)
Comment (vpatakottu) [reply test/Microsoft.Sbom.Targets.Tests/AbstractGenerateSBomTaskInputTests.cs]: Updated!

**Review by gustavoaca1997 — APPROVED** [review]

**Pull request merged**